Automattic/jetpack
# Build Action: post reminder comment to test on WordPress.com

#### Changes proposed in this Pull Request:

This introduces 2 new steps to the Build action:

- In the beginning of the build run, it checks if we need to post a test reminder comment on the PR.
    - If one is already posted, we do not need to post another.
    - If none exists, we create such a comment, with some generic instructions.
    - If one is already posted but the PR does not touch Jetpack any longer (i.e. does not touch de-Fusioned files), we delete the comment.
- Once the build is complete, the comment, if it exists, is updated with instructions on how to test.

~This posts a new comment. I think that is fine, but part of me wonders if we should instead edit the big comment that gets posted by the Repo gardening action.~ This was discussed below: https://github.com/Automattic/jetpack/pull/25310#pullrequestreview-1054416035

> **Note**
> Following the discussion at p9dueE-5yg-p2, I've introduced a new file, `.github/actions/build/de-fusioned-files.js`, where we'll keep a list of all de-fusioned files as we go through the project. We'll use that list to figure out if a test reminder comment needs to be posted on PRs. When Fusion is no more, we can retire that list, and instead rely on the output of the `changed` step just like we did early on this PR.

> **Note**
> You'll note some debug statements in the code. Those will not be displayed in the action output by default, but can be displayed if one enables debugging mode for GitHub actions as explained [here](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging). Here is how the debug statements can look like:

<img width="962" alt="image" src="https://user-images.githubusercontent.com/426388/183912252-ee68c01d-bfe5-4d37-980e-c67b6a81096c.png">

**To-do**

- [x] If we go with this approach, we would probably want to limit the comment to PRs that touch the Jetpack plugin. I think we may be able to achieve that by moving the step that generates the first comment a bit later in the Build process, after the "Detect changed projects" step, so it can get the projects that are touched, and only comment if Jetpack is concerned.
- [x] We need to ensure that the comment only gets posted once, even if the build workflow is triggered multiple times (like when someone pushes to the existing branch). A quick and dirty solution would be to check for the `opened` event type in addition to the `pull_request` event name before we post the comment, but this would not support folks that open a PR with no changes to the Jetpack plugin, and then later on update their PR with new changes that touch the plugin. 

#### Other information:


#### Jetpack product discussion

Internal reference: p1658931404733059-slack-C03QNBQKG73

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

* Here is how the comment below looked like immediately after opening the PR:

<img width="967" alt="image" src="https://user-images.githubusercontent.com/426388/181575588-db016129-3124-431e-9df2-f991d1921414.png">

* Here is how it looks once the build has completed:

<img width="988" alt="image" src="https://user-images.githubusercontent.com/426388/181579042-21f7b937-c452-4fbf-8122-0937ff58f8e7.png">

You can see the tests here:
- A comment is posted when a de-fusioned file (`projects/plugins/jetpack/_inc/genericons.php`) is edited: #25313
- No comment is posted when the PR touches a Fusioned file: #25314

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - 0-as-1202637881054088
